helix-core/src/doc_formatter.rs

```diff
@@ -0,0 +1,340 @@
+//! The `DocumentFormatter` forms the bridge between the raw document text
+//! and onscreen positioning. It yields the text graphemes as an iterator
+//! and transverses (part) of the document text. During that transversal it
+//! handles grapheme detection, softwrapping and annotation.
+//! It yields `FormattedGrapheme`s and their corresponding visual coortdinates.
+//!
+//! As both virtual text and softwrapping can insert additional lines into the document
+//! it is generally not possible to find the start of the previous visual line.
+//! Instead the `DocumentFormatter` starts at the last "checkpoint" (usually a linebreak)
+//! called a "block" and the caller must advance it as needed.
+
+use std::borrow::Cow;
+use std::fmt::Debug;
+use std::mem::{replace, take};
+
+#[cfg(test)]
+mod test;
+
+use unicode_segmentation::{Graphemes, UnicodeSegmentation};
+
+use crate::graphemes::Grapheme;
+use crate::syntax::Highlight;
+use crate::text_annotations::TextAnnotations;
+use crate::{Position, RopeGraphemes, RopeSlice};
+
+/// A preprossed Grapheme that is ready for rendering
+/// with attachted styling data
+#[derive(Debug, Clone)]
+pub struct FormattedGrapheme<'a> {
+    pub grapheme: Grapheme<'a>,
+    pub highlight: Option<Highlight>,
+    // the number of chars in the document required by this grapheme
+    pub doc_chars: u16,
+}
+
+impl<'a> FormattedGrapheme<'a> {
+    /// Returns whether this grapheme is virtual inline text
+    pub fn is_virtual(&self) -> bool {
+        // The highlight field is only used for inline virtual text
+        // so it's save to reuse that.
+        // We can not use doc_chars here as that is also 0 for the EOF space
+        let is_virtual = self.highlight.is_some();
+        if is_virtual {
```

Comment:
Minor: Should the whole `if` block be conditionally compiled?

Reply:
code doesn't exist anymore with latest refactor